### ISSUE TITLE
DRA: detect stale DRA plugin sockets

### ISSUE THREAD

**Issue by pohly** [kubernetes/kubernetes#128696]
### What would you like to be added?

@klueska observed that after uninstalling his DRA driver *without* removing the Unix Domain socket used for gRPC towards the kubelet, kubelet didn't unregister the driver. We may have to add some liveness probing to `pkg/kubelet/cm/dra/plugin/registration.go`.

/sign node
/wg device-management
/priority backlog

### Why is this needed?

The kubelet needs to remove ResourceSlices on behalf of a defunct DRA driver.

**Comment by pohly** [comment]
A good first step would be to write an E2E test where removing the test driver doesn't remove the socket.

**Comment by ravisantoshgudimetla** [comment]
Struggling with same thing as well. 

In general, I'm in agreement with what was proposed [here](https://github.com/kubernetes/kubernetes/issues/128696#issuecomment-2465268252) but do we want to support it in tree? Do we see 
upgrade == unregistering old version of plugin + registering latest version of plugin to keep things simple and immutable?
or store the version of plugin in metadata and do an incremental version check?

**Comment by pohly** [comment]
Under which circumstances do we have two different versions of a driver running at the same time? Is that intentional?

If a driver gets deployed as a daemonset, I would expect that there is always only one pod running. There is downtime as the old pod shuts down and the new one starts up. Or is it more seamless and both run in parallel? I haven't tried it.

**Comment by linxiulei** [comment]
> There is downtime as the old pod shuts down and the new one starts up. Or is it more seamless and both run in parallel? I haven't tried it.

The idea to have two Pods running in parallel is to eliminate the downtime so yes, it will become more seamless if that's supported.

Currently if two device plugin pods are running ini parallel, it only results in confusion and race condition to kubelet

**Comment by bart0sh** [comment]
/assign
@pohly Feel free to re-assign if you're going to work on this issue as part of https://github.com/kubernetes/kubernetes/issues/128964

**Comment by tico88612** [comment]
Hello @pohly @bart0sh 
This issue and related PR seems to be under review. If there's anything we can do, please let us know.
The code freeze starts at 02:00 UTC on Friday, 21st March 2025 (about 5 weeks from now). Please make sure the PR has both lgtm and approved labels before the code freeze. Thanks!

**Comment by tico88612** [comment]
Hello @pohly @SergeyKanzhelev @bart0sh 
Appreciate all of your efforts with this Issue! Is the plan still to resolve it for v1.33 release?
If so, a gentle reminder that the code freeze has started [02:00 UTC Friday 21st March 2025](https://everytimezone.com/s/2c5e9275) . Please make sure any PRs have both lgtm and approved labels ASAP, and file an [Exception](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md) if you haven't done it yet.
Thanks!

**Comment by liggitt** [comment]
this is still in the 1.33 milestone and has a pending PR ... is it actually targeting 1.33 still?

**Comment by pohly** [comment]
No.

/milestone clear

**Comment by bart0sh** [comment]
@pohly Can you reopen this issue? We implemented connection monitoring, but it looks like the issue is not fully fixed yet. @nojnhuh [reproduced it](https://github.com/kubernetes-sigs/dra-example-driver/pull/104/files#r2170136676).

**Comment by pohly** [comment]
/reopen
/unassign
/assign @bart0sh 

To investigate https://github.com/kubernetes-sigs/dra-example-driver/pull/104/files#r2170136676.

**Comment by nojnhuh** [comment]
> [@pohly](https://github.com/pohly) Can you reopen this issue? We implemented connection monitoring, but it looks like the issue is not fully fixed yet. [@nojnhuh](https://github.com/nojnhuh) [reproduced it](https://github.com/kubernetes-sigs/dra-example-driver/pull/104/files#r2170136676).

In addition to that thread, one extra bit of context I've noticed is that the `connection refused` errors logged by kubelet seem to stop after 30m. It looks like there was some exponential backoff occurring too where the last two requests were 2m14s apart and the backoff still hadn't hit its max interval.

**Comment by bart0sh** [comment]
I'm going to close this issue again as I believe it's fixed by https://github.com/kubernetes/kubernetes/pull/132058. 
Kubelet can't delete stale sockets as this can race with drivers. Drivers should take care of this in my opinion.
However, DRA manager(part of Kubelet) detects socket disconnects and deletes driver's resource slices, which prevents  scheduler from scheduling new pods to the nodes with non-functonal DRA drivers. That's all we can do on the Kubelet side.

https://github.com/kubernetes/kubernetes/issues/132660 is another issue and needs to be fixed separately.
